crates/biome_js_analyze/src/lint/nursery/no_shadow.rs

```diff
@@ -0,0 +1,227 @@
+use biome_analyze::{
+    QueryMatch, Rule, RuleDiagnostic, RuleSource, context::RuleContext, declare_lint_rule,
+};
+use biome_console::markup;
+use biome_js_semantic::{Binding, SemanticModel};
+use biome_js_syntax::{
+    JsClassExpression, JsFunctionExpression, JsIdentifierBinding, JsVariableDeclarator,
+    TsIdentifierBinding, TsTypeAliasDeclaration,
+};
+use biome_rowan::{AstNode, SyntaxNodeCast, TokenText, declare_node_union};
+
+use crate::services::semantic::SemanticServices;
+
+declare_lint_rule! {
+    /// Disallow variable declarations from shadowing variables declared in the outer scope.
+    ///
+    /// Shadowing is the process by which a local variable shares the same name as a variable in its containing scope. This can cause confusion while reading the code and make it impossible to access the global variable.
+    ///
+    /// See also: [`noShadowRestrictedNames`](http://biome.dev/linter/rules/no-shadow-restricted-names)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "bar";
+    /// if (true) {
+    ///    const foo = "baz";
+    /// }
+    /// ```
+    ///
+    /// Variable declarations in functions can shadow variables in the outer scope:
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "bar";
+    /// const bar = function () {
+    ///     const foo = 10;
+    /// }
+    /// ```
+    ///
+    /// Function argument names can shadow variables in the outer scope:
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "bar";
+    /// function bar(foo) {
+    ///     foo = 10;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const foo = "bar";
+    /// if (true) {
+    ///    const qux = "baz";
+    /// }
+    /// ```
+    ///
+    pub NoShadow {
+        version: "next",
+        name: "noShadow",
+        language: "js",
+        recommended: false,
+        sources: &[
+            RuleSource::Eslint("no-shadow"),
+            // uncomment when we can handle the test cases from typescript-eslint
+            // RuleSource::EslintTypeScript("no-shadow"),
+        ],
```

Comment:
I have excluded the unit tests from `typescript-eslint`, and therefore the rule source here, because of #5822.